d2/src/main/java/com/linkedin/d2/balancer/clusterfailout/ClusterFailoutBucketConfig.java

```diff
@@ -0,0 +1,129 @@
+package com.linkedin.d2.balancer.clusterfailout;
+
+import java.util.Map;
+import java.util.Date;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class is a simple data structure for tracking the traffic state of a given application.
+ */
+
+
+/*
+failedoutBuckets: array[BucketFailoutConfig] {
+    fabric: FabricUrn
+    partition: string
+    // e.g. buckets 0-4 are offline
+    bucketIdRange: IntRange
+    offlineAt: Time
+    onlineAt: optional Time
+  }
+ */
+public class ClusterFailoutBucketConfig
```

Comment:
FabricUrn, partition, buckets all seem to be quite LI specific, and i'm a little concerned of these concepts leak into open-source projects. Will all of these needed by D2? I thought D2 only needs to know which fabric is being failed out.

Reply:
Fabric is needed to indicate which fabric this is referring to.
Partition is needed to indicate which traffic partition this is impacting (As we have multiple traffic partitions).
Bucket is our grouping of which sets of traffic to fail out.

I agree that all of these are fairly LI specific; We do need all of them though. Do you have a suggestion for how to implement these in a more generic manner?

Reply:
In this case, we may need to keep them as generic `Map<String, Object>` and have the d2 client delegator to parse the data. In addition, we will need to have the d2 client delegator to register callback/listeners with the metadata provider so that it can get notified that it needs to update/parse new config.
+@jpstewart wdyt?

Reply:
We could have more general field names (like "fabric" instead of fabricUrn, cause Urn is LI specific). The data types are already general (String, Integer, Date, etc.), Object type will be a bit overkill I think. The d2 client delegator can parse the String into fabricUrn when needed.

Reply:
In this case, maybe we are better off to just store `fabric` in the config? Since we also need to setup watches on peer clusters and need to construct the cluster name by appending the fabric suffix.
fwiw, [URN](https://en.wikipedia.org/wiki/Uniform_Resource_Name) is not LI specific though.

Reply:
Can this be abstracted into an interface? The value of the bucket triple can be abstracted behind that, and the d2 delegate implementation can assume the proper implementation of the interface?

Reply:
This is my WIP diff: https://github.com/linkedin/rest.li/commit/c40f4c8360678626453654183fd84dafb53da067

In summary, we will have an abstract `ClusterFailoutConfigProvider` defined in `rest.li`. It will have an abstract method to convert `ClusterFailoutProperties` to `ClusterFailoutConfig`. A concrete implementation of `ClusterFailoutConfigProvider` will be created in `service-level-failout` mp and `container` will use factories to pull in and hook it into the `D2Client` builders.

Reply:
Made these types generic as per our conversation.